unix/modjni.c

```diff
@@ -0,0 +1,375 @@
+
+    const char *name = qstr_str(self->name);
+//    jstring meth_name = JJ(NewStringUTF, name);
+
+    jclass obj_class = JJ(GetObjectClass, self->obj);
+    jarray methods = JJ(CallObjectMethod, obj_class, Class_getMethods_mid);
+
+    return call_method(self->obj, name, methods, false, n_args, args);
+}
+
+STATIC const mp_obj_type_t jmethod_type = {
+    { &mp_type_type },
+    .name = MP_QSTR_jmethod,
+    .print = jmethod_print,
+    .call = jmethod_call,
+//    .attr = jobject_attr,
+//    .locals_dict = (mp_obj_t)&jobject_locals_dict,
+};
+
+#ifdef __ANDROID__
+#define LIBJVM_SO "libdvm.so"
+#else
+#define LIBJVM_SO "libjvm.so"
+#endif
+
+STATIC void create_jvm() {
+    JavaVMInitArgs args;
+    JavaVMOption options;
+    options.optionString = "-Djava.class.path=.";
+    args.version = JNI_VERSION_1_6;
+    args.nOptions = 1;
+    args.options = &options;
+    args.ignoreUnrecognized = 0;
+
+    if (env) {
+        return;
+    }
+
+    void *libjvm = dlopen(LIBJVM_SO, RTLD_NOW | RTLD_GLOBAL);
+    if (!libjvm) {
+        nlr_raise(mp_obj_new_exception_msg_varg(&mp_type_OSError, "unable to load libjvm.so, use LD_LIBRARY_PATH"));
+    }
+    int (*_JNI_CreateJavaVM)(void*, void**, void*) = dlsym(libjvm, "JNI_CreateJavaVM");
+
+    int st = _JNI_CreateJavaVM(&jvm, (void**)&env, &args);
+    if (st < 0 || !env) {
+        nlr_raise(mp_obj_new_exception_msg_varg(&mp_type_OSError, "unable to create JVM"));
+    }
+
+    jclass class_class = JJ(FindClass, "java/lang/Class");
+    jclass method_class = JJ(FindClass, "java/lang/reflect/Method");
+    String_class = JJ(FindClass, "java/lang/String");
+
+    Class_getField_mid = (*env)->GetMethodID(env, class_class, "getField",
+                                     "(Ljava/lang/String;)Ljava/lang/reflect/Field;");
+    Class_getMethods_mid = (*env)->GetMethodID(env, class_class, "getMethods",
+                                     "()[Ljava/lang/reflect/Method;");
+    Class_getConstructors_mid = (*env)->GetMethodID(env, class_class, "getConstructors",
+                                     "()[Ljava/lang/reflect/Constructor;");
+    Method_getName_mid = (*env)->GetMethodID(env, method_class, "getName",
+                                     "()Ljava/lang/String;");
+    Method_toString_mid = (*env)->GetMethodID(env, method_class, "toString",
+                                     "()Ljava/lang/String;");
+}
+
+STATIC mp_obj_t mod_jni_cls(mp_obj_t cls_name_in) {
+    const char *cls_name = mp_obj_str_get_str(cls_name_in);
+    if (!env) {
+        create_jvm();
+    }
+    jclass cls = JJ(FindClass, cls_name);
+
+    mp_obj_jclass_t *o = m_new_obj(mp_obj_jclass_t);
+    o->base.type = &jclass_type;
+    o->cls = cls;
+    return o;
+}
+MP_DEFINE_CONST_FUN_OBJ_1(mod_jni_cls_obj, mod_jni_cls);
+
+STATIC const mp_map_elem_t mp_module_jni_globals_table[] = {
+    { MP_OBJ_NEW_QSTR(MP_QSTR___name__), MP_OBJ_NEW_QSTR(MP_QSTR_jni) },
+    { MP_OBJ_NEW_QSTR(MP_QSTR_cls), (mp_obj_t)&mod_jni_cls_obj },
```

Comment:
Pretty much, yes. Except that `cls` requires native JNI class name (slash-separated, not dot-separated).